### ISSUE TITLE
[FEATURE] Adding Carlini and Wagner's L0 Attack

### ISSUE THREAD

**Issue by mbeliaev1** [Harry24k/adversarial-attacks-pytorch#98]
### ✨ Short description of the feature [tl;dr]

Is it possible to include Carlini and Wagner's L0 attack strategy to this repository?

### 💬 Detailed motivation and codes

I see that [Carlini & Wagner's](https://arxiv.org/pdf/1608.04644.pdf) L2 attack is included in the repository already.  The same paper introduced an L0 attack strategy which is an improvement over JSMA. Since both these attacks came from the same repository, would it be possible to add this attack? If this was already tried, was there any reason for not including it? 

I may go about doing this myself as the original code is in tensorflow and I would like to move it over to torch.

**Comment by Harry24k** [comment]
Hello!

First, to my best knowledge, the CW L0 attack ( and also JSMA) is not widely used for adversarial attacks and adversarial defenses compared to other attacks such as the OnePixel and Sparse attacks.

For this reason, I've never tried implementation of the CW L0 and JSMA attacks. However, if it is possible to implement, it would be really awesome :).

**Comment by mbeliaev1** [comment]
Thanks for your reply!

Yes, it is true L0 attacks are in general not so common in the literature due to their complex geometry and non-convex nature. 
A lot of defenses for L0 end up including non-differentiable components, and hence sparse "black-box" attacks are usually the go to, at least from my understanding. 

This brings to mind two more attacks that I see in the literature (and which have torch implementations): 
-  Pointwise Attack from Schott et al. which is part of the foolbox

**Comment by mbeliaev1** [comment]
Thanks for your reply!

Yeah I agree, in general L0 attacks are less popular due to their non-convex geometry, and defenses designed for them typically have components which are not differentiable making sparse black-box attacks more common. 

If I go ahead and implement CW-0 for my own purposes, I will let you know. I plan to use your repo anyway so I will try and keep it streamlined. 

Since I am on the topic, are you aware of these other two attack methods for L-0 which are both from published papers?

- [sparse-rs](https://github.com/fra31/sparse-rs) a black box random search attack
- [PGD-0](https://github.com/fra31/sparse-imperceivable-attacks) gradient based l0 attack

You can close the issue btw, I am just curious if you had any experience implementing them here and maybe chose not to include.
